materialize-bigquery/schema.go

```diff
@@ -0,0 +1,292 @@
+		// When a projection is a string and it has the Inference_String set, it's possible for the user
+		// to pass along more information about the string field to let the connector create a more
+		// appropriate field for that string. If an Inference_String{} is present, it doesn't default back
+		// to a normal string as it is prefered to let the user know that the inference could not be matched.
+		case "string":
+			if projection.Inference.String_ != nil {
+				s := projection.Inference.String_
+				switch s.Format {
+				case "date-time":
+					f.fieldType = bigquery.TimestampFieldType
+				case "date":
+					f.fieldType = bigquery.DateFieldType
+				// There is a case where a String_ is present but the values aren't set.
+				// I'm not sure how this can happen since the protobuf struct should
+				// be optional and nil, but it did happen in my testing and I couldn't
+				// figured out why, so this is here as a precaution (P-O)
+				case "":
+					f.fieldType = bigquery.StringFieldType
+					f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer)
+
+				default:
+					return nil, fmt.Errorf("inferring string column type as the format is not supported: %s", s.Format)
+				}
+			} else {
+				f.fieldType = bigquery.StringFieldType
+				f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer)
+			}
+		case "boolean":
+			f.fieldType = bigquery.BooleanFieldType
+		case "integer":
+			f.fieldType = bigquery.IntegerFieldType
+		case "number":
+			f.fieldType = bigquery.BigNumericFieldType
+		case "object":
+			f.fieldType = bigquery.StringFieldType
+			f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer, sqlDriver.SingleQuotesWrapper())
+		case "array":
+			f.fieldType = bigquery.StringFieldType
+			f.sanitizers = append(f.sanitizers, sqlDriver.DefaultQuoteSanitizer, sqlDriver.SingleQuotesWrapper())
+		}
+	}
+
+	if f.fieldType == "" {
+		return nil, fmt.Errorf("could not map the field to a big query type: %s", projection.Field)
+	}
+
+	return f, nil
+}
+
+func (f *Field) FieldType() bigquery.FieldType {
+	return bigquery.FieldType(f.fieldType)
+}
+
+func (f *Field) Name() string {
+	return f.name
+}
+
+func (f *Field) Render(val interface{}) (str string, err error) {
+	str = fmt.Sprint(val)
+
+	for _, sanitizer := range f.sanitizers {
+		str = sanitizer(str)
+	}
+
+	return str, err
+}
+
+func (f *Field) FieldSchema(projection *pf.Projection) (*bigquery.FieldSchema, error) {
+	var includesNull bool
+
+	// This is a test to make sure that their is only 1 infered type for this projection
+	// ie. Inference.Types = [string, integer] is wrong. [string, null] is OK. [string, integer, null] is wrong
+	if !projection.Inference.IsSingleType() {
+		return nil, fmt.Errorf("don't expect multiple types besides Null and a real type, can't proceed: %s", &projection.Inference.Types)
+	}
+
+	for _, p := range projection.Inference.Types {
+		if p == "null" {
+			includesNull = true
```

Comment:
nit: after the above check for `IsSingleType`, if `p == "null"`, then this field must _always_ be null.